worker/restore_reduce.go

```diff
@@ -0,0 +1,314 @@
+
+	x.Check2(io.ReadFull(reader, headerBuf))
+	header := &pb.MapHeader{}
+	err = header.Unmarshal(headerBuf)
+	x.Check(err)
+
+	itr := &mapIterator{
+		fd:     fd,
+		reader: reader,
+	}
+	return header, itr
+}
+
+type reducer struct {
+	mapDir        string
+	mapItrs       []*mapIterator
+	partitionKeys [][]byte
+	bufferCh      chan *z.Buffer
+	w             Writer
+
+	bytesProcessed uint64
+	bytesRead      uint64
+}
+
+type Writer interface {
+	Write(buf *z.Buffer) error
+}
+
+func RunReducer(w Writer, mapDir string) error {
+	r := &reducer{
+		w:        w,
+		bufferCh: make(chan *z.Buffer, 10),
+		mapDir:   mapDir,
+	}
+	closer := z.NewCloser(1)
+	defer closer.SignalAndWait()
+	go r.Progress(closer)
+
+	return r.Reduce()
+}
+
+func (r *reducer) Progress(closer *z.Closer) {
+	defer closer.Done()
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	start := time.Now()
+	update := func() {
+		since := time.Since(start)
+		read := atomic.LoadUint64(&r.bytesRead)
+		proc := atomic.LoadUint64(&r.bytesProcessed)
+		pr := uint64(float64(proc) / since.Seconds())
+		glog.Infof(
+			"Restore REDUCE %s read: %s. processed: %s. rate: %s/sec. jemalloc: %s.\n",
+			x.FixedDuration(since), humanize.IBytes(read), humanize.IBytes(proc),
+			humanize.IBytes(pr), humanize.IBytes(uint64(z.NumAllocBytes())))
+	}
+	for {
+		select {
+		case <-closer.HasBeenClosed():
+			update()
+			glog.Infof("Restore REDUCE Done in %s.\n", x.FixedDuration(time.Since(start)))
+			return
+		case <-ticker.C:
+			update()
+		}
+	}
+}
+
+func (r *reducer) Reduce() error {
+	var files []string
+
+	var total int64
+	f := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if strings.HasSuffix(info.Name(), ".map") {
+			files = append(files, path)
+			total += info.Size()
+		}
+		return nil
+	}
+
+	if err := filepath.Walk(r.mapDir, f); err != nil {
+		return err
+	}
+	glog.Infof("Got %d map files of compressed size: %s.\n",
+		len(files), humanize.IBytes(uint64(total)))
+
+	// Pick up map iterators and partition keys.
+	partitions := make(map[string]struct{})
+	for _, fname := range files {
+		header, itr := newMapIterator(fname)
+		for _, k := range header.PartitionKeys {
+			if len(k) == 0 {
+				continue
+			}
+			partitions[string(k)] = struct{}{}
+		}
+		r.mapItrs = append(r.mapItrs, itr)
+	}
+
+	keys := make([][]byte, 0, len(partitions))
+	for k := range partitions {
+		keys = append(keys, []byte(k))
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return y.CompareKeys(keys[i], keys[j]) < 0
+	})
+	// Append nil for the last entries.
+	keys = append(keys, nil)
+	r.partitionKeys = keys
+
+	errCh := make(chan error, 2)
+	go func() {
+		errCh <- r.blockingRead()
+	}()
+	go func() {
+		errCh <- r.process()
+	}()
+
+	for i := 0; i < 2; i++ {
+		if err := <-errCh; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *reducer) blockingRead() error {
+	cbuf := z.NewBuffer(64<<20, "Restore.GetBuf")
```

Comment:
Nit: Should we use constants?

Reply:
This whole needs to be figured out better.